parachain/src/wasm_executor/validation_host/workspace.rs

```diff
@@ -0,0 +1,556 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This module implements a "workspace" - basically a wrapper around a shared memory that
+//! is used as an IPC channel for communication between the validation host and it's validation
+//! worker.
+
+use crate::primitives::{ValidationParams, ValidationResult};
+use super::LOG_TARGET;
+use parity_scale_codec::{Decode, Encode};
+use raw_sync::{
+	events::{Event, EventImpl, EventInit, EventState},
+	Timeout,
+};
+use shared_memory::{Shmem, ShmemConf};
+use std::{
+	error::Error,
+	fmt,
+	io::{Cursor, Write},
+	slice,
+	sync::atomic::AtomicBool,
+	time::Duration,
+};
+
+// maximum memory in bytes
+const MAX_PARAMS_MEM: usize = 1024 * 1024; // 1 MiB
```

Comment:
Is 1 Mib hard limit on proof size really enough?

Reply:
Yeah, good point. We over allocate the buffer so it will still fit, but this could be express more clearer.

Reply:
By how much do we over allocate? Or do you mean that the POV could be any size and we always can allocate enough memory for it?

Reply:
Scratch that, even if we over allocate, the call data is rejected by the params size check. See the fix here https://github.com/paritytech/polkadot/pull/2445